master/vol.go

```diff
@@ -456,6 +456,60 @@ func (vol *Vol) getDataPartitionByID(partitionID uint64) (dp *DataPartition, err
 	return vol.dataPartitions.get(partitionID)
 }
 
+func (vol *Vol) addMetaPartitions(c *Cluster, count int) (err error) {
+	// add extra meta partitions at a time
+	var (
+		start uint64
+		end   uint64
+	)
+
+	vol.createMpMutex.Lock()
+	defer vol.createMpMutex.Unlock()
+
+	// update End of the maxMetaPartition range
+	maxPartitionId := vol.maxPartitionID()
+	rearMetaPartition := vol.MetaPartitions[maxPartitionId]
+	oldEnd := rearMetaPartition.End
+	end = rearMetaPartition.MaxInodeID + gConfig.MetaPartitionInodeIdStep
+
+	if err = rearMetaPartition.canSplit(end, gConfig.MetaPartitionInodeIdStep, false); err != nil {
+		return err
+	}
+
+	rearMetaPartition.End = end
+	if err = c.syncUpdateMetaPartition(rearMetaPartition); err != nil {
+		rearMetaPartition.End = oldEnd
+		log.LogErrorf("action[addMetaPartitions] split partition partitionID[%v] err[%v]", rearMetaPartition.PartitionID, err)
+		return
+	}
+
+	// create new meta partitions
+	for i := 0; i < count; i++ {
+		start = end + 1
+		end = start + gConfig.MetaPartitionInodeIdStep
+
+		if end > (defaultMaxMetaPartitionInodeID - gConfig.MetaPartitionInodeIdStep) {
+			end = defaultMaxMetaPartitionInodeID
+			log.LogWarnf("action[addMetaPartitions] vol[%v] add too many meta partition ,partition range overflow ! ", vol.Name)
+		}
+
+		if i == count-1 {
+			end = defaultMaxMetaPartitionInodeID
+		}
+
+		if err = vol.createMetaPartition(c, start, end); err != nil {
```

Comment:
Does this logic conflict with the logic of automatic creation? I have doubts about it, and it may not be your fault. I would appreciate it if you could help analyze it. Thank you.

Reply:
Before call createMetaPartition which based on doCreateMetaPartition , i acquired createMpMutex .  
And any mp auto creation task will  go through splitMetaPartition → doSplitMetaPartition → doCreateMetaPartition. 
In splitMetaPartition it also try to acquire createMpMutex, and when acquired  it double check the maxPartitionID, if the maxPartitionID changed , the task will abort and retry.  So i think the logic is safe
```go
  vol.createMpMutex.Lock()
	defer vol.createMpMutex.Unlock()

	maxPartitionID := vol.maxPartitionID()
	if maxPartitionID != mp.PartitionID {
		err = fmt.Errorf("mp[%v] is not the last meta partition[%v]", mp.PartitionID, maxPartitionID)
		return
	}
```